framework/sandstone_tests.cpp

```diff
@@ -0,0 +1,255 @@
+{
+    std::vector<std::string> tokens;
+    auto it = line.begin();
+    while (is_ignored(*it)) ++it;
+    if (*it == '#' || it == line.end()) return LT_EMPTY;
+    for (; ; ++it) {
+        auto cit = it;
+        /* scroll while it's valid token contents: till end of line, a
+         * terminator, or a space. */
+        while (cit != line.end() && !is_terminator(*cit) && !is_ignored(*cit)) ++cit;
+        auto tend = cit;
+        while (cit != line.end() && is_ignored(*cit)) ++cit;
+        if (cit != line.end() && !is_terminator(*cit)) return LT_SYNTAX_ERROR;
+        tokens.emplace_back(it, tend);
+        if (cit == line.end() || *cit == '#') break;
+        it = cit;
+    }
+    if (!tokens.size()) __builtin_unreachable();
+    if (tokens.size() > 2) return LT_SYNTAX_ERROR;
+    SandstoneTestSet::TestSet set = test_set->lookup(tokens[0].c_str());
+    if (!set.size()) return LT_TEST_NOT_FOUND;
+    if (set.size() != 1) return LT_SYNTAX_ERROR; /* Artificially do not allow specifying wildcards or groups in the list file. */
+    ti.test = set[0];
+    if (tokens.size() == 2) ti.attribute = tokens[1];
+    return LT_VALID_TEST;
+}
+
+static std::vector<struct test_cfg_info> load_test_list(std::ifstream &fstream, SandstoneTestSet *test_set, bool ignore_unknown_tests, std::vector<std::string> &errors)
+{
+    std::vector<struct test_cfg_info> res;
+    std::string line;
+    bool error = false;
+    int lineno = 1;
+    while (std::getline(fstream, line)) {
+        struct test_cfg_info ti;
+        switch (parse_test_list_line(line, test_set, ti)) {
+            case LT_VALID_TEST:
+                res.push_back(ti);
+                break;
+            case LT_SYNTAX_ERROR:
+            {
+                std::stringstream msg_strm;
+                msg_strm <<  "Syntax error: line " << lineno << ": " << line;
+                errors.push_back(msg_strm.str());
+                error = true;
+                break;
+            }
+            case LT_EMPTY:
+                break;
+            case LT_TEST_NOT_FOUND:
+                if (!ignore_unknown_tests) {
+                    std::stringstream msg_strm;
+                    msg_strm << "Unknown test:  " << line;
+                    errors.push_back(msg_strm.str());
+                    error = true;
+                }
+                break;
+        }
+        ++lineno;
+    }
+    if (error) {
+        res.clear();
+    }
+    return res;
+}
+
+std::vector<struct test_cfg_info> SandstoneTestSet::add_test_list(const char *fname, std::vector<std::string> &errors)
+{
+    std::ifstream list_file(fname, std::ios_base::in);
+    std::vector<struct test_cfg_info> entries = load_test_list(list_file, this, cfg.ignore_unknown_tests, errors);
+    if (!errors.empty()) return {};
+    for (auto e : entries) {
+        if (!e.attribute.empty()
+                && strcasecmp(e.attribute.c_str(), "default")) {
+            e.test->desired_duration = string_to_millisecs(e.attribute).count();
```

Comment:
I've just debugged a buggy test that ran into an infinite loop when its `desired_duration` field was changed from -1 to a positive value. Can you fix this here so we never reset a -1?

Reply:
Done. See the last commit.